modules/fedimint-wallet-client/src/deposit.rs

```diff
@@ -0,0 +1,368 @@
+                        .output
+                        .iter()
+                        .enumerate()
+                        .find_map(|(idx, output)| {
+                            if output.script_pubkey == script {
+                                Some(idx as u32)
+                            } else {
+                                None
+                            }
+                        })
+                        .expect("TODO: handle invalid tx returned by API");
+
+                    return (transaction, out_idx);
+                } else {
+                    trace!("No transactions received yet for script {script:?}");
+                }
+            }
+            Err(e) => {
+                warn!("Error fetching transaction history for {script:?}: {e}");
+            }
+        }
+
+        sleep(Duration::from_secs(10)).await;
+    }
+}
+
+async fn get_script_history(
+    esplora_server: &str,
+    script: &Script,
+) -> anyhow::Result<Vec<bitcoin::Transaction>> {
+    let esplora_client = esplora_client(esplora_server)?;
+    let script = script.clone();
+
+    let transactions = esplora_client
+        .scripthash_txs(&script, None)
+        .await?
+        .into_iter()
+        .map(|tx| tx.to_tx())
+        .collect::<Vec<_>>();
+
+    Ok(transactions)
+}
+
+async fn transition_tx_seen(
+    old_state: DepositStateMachine,
+    btc_transaction: bitcoin::Transaction,
+    out_idx: u32,
+) -> DepositStateMachine {
+    let DepositStateMachine {
+        operation_id,
+        state: old_state,
+    } = old_state;
+
+    match old_state {
+        DepositStates::Created(created_state) => DepositStateMachine {
+            operation_id,
+            state: DepositStates::WaitingForConfirmations(WaitingForConfirmationsDepositState {
+                tweak_key: created_state.tweak_key,
+                btc_transaction,
+                out_idx,
+            }),
+        },
+        state => panic!("Invalid previous state: {state:?}"),
+    }
+}
+
+async fn await_deposit_address_timeout(timeout_at: SystemTime) {
+    if let Ok(time_until_deadline) = timeout_at.duration_since(fedimint_core::time::now()) {
+        tokio::time::sleep(time_until_deadline).await;
+    }
+}
+
+async fn transition_deposit_timeout(old_state: DepositStateMachine) -> DepositStateMachine {
+    assert!(
+        matches!(old_state.state, DepositStates::Created(_)),
+        "Invalid previous state"
+    );
+
+    DepositStateMachine {
+        operation_id: old_state.operation_id,
+        state: DepositStates::TimedOut(TimedOutDepositState {}),
+    }
+}
+
+async fn await_btc_transaction_confirmed(
+    context: WalletClientContext,
+    global_context: DynGlobalClientContext,
+    waiting_state: WaitingForConfirmationsDepositState,
+) -> TxOutProof {
+    loop {
+        // Wait for confirmation
+        let consensus_height = match global_context
+            .api()
+            .with_module(LEGACY_HARDCODED_INSTANCE_ID_WALLET)
```

Comment:
To get rid of this would we need to pass in the module instance id from the `WalletClient`?

Reply:
It's probably better to give the LN module its own API endpoint longer term.